src/System.Management.Automation/engine/parser/Compiler.cs

```diff
@@ -955,33 +955,95 @@ private IEnumerable<Expression> CompileInvocationArguments(IEnumerable<Expressio
             return arguments == null ? Array.Empty<Expression>() : arguments.Select(CompileExpressionOperand);
         }
 
-        internal Expression ReduceAssignment(ISupportsAssignment left, TokenKind tokenKind, Expression right)
+        internal Expression ReduceAssignment(
+            ISupportsAssignment left,
+            TokenKind tokenKind,
+            Expression right,
+            MemberExpressionAst lhsMemberAccessAst)
```

Comment:
Looking at how this is called, I think it does make sense the way it's currently written, but I just want to raise it for discussion.

A new possibility in one of the callers in this method now forces itself on all the other callers, leading to a number of calls where the caller must pass `null` in here. This seems to be common in the PowerShell codebase, but in my opinion leads to decohesion of methods, both from the perspective of the caller (which must now carefully select the appropriate subset of parameters) and from the perspective of the method itself (which must now carefully step between shared and scenario-specific logic).

In some respects this new parameter is a pseudo-boolean; it's either null or has a particular value that causes us to follow another path within the method.

That suggests to me that really we have two methods on our hands, with some shared logic best represented by breaking things up into smaller, composable methods.

Reply:
Definitely a good point. I can have a look at this and try to refactor things a bit.

Reply:
I think I have a pretty good way to do this. There's _some_ duplication of code, but it separates the code paths much more neatly. I'll commit this change shortly; any additional suggestions you have to improve what I managed to get to are more than welcome! 😊